File: app/src/main/java/com/gtlx/statusbardrift/MainActivity.java
package com.gtlx.statusbardrift;

import android.app.Activity;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.SeekBar;
import android.widget.ScrollView;
import android.widget.Button;
import android.graphics.Color;
import android.view.Gravity;
import android.util.TypedValue;

/**
 * 状态栏漂流瓶 —— 设置界面
 */
public class MainActivity extends Activity {

    private static final String PREFS_NAME = "status_bar_drift";
    private static final String KEY_DRIFT_PX = "drift_px";
    private static final String KEY_INTERVAL_SEC = "interval_sec";

    private static final int DEFAULT_DRIFT_PX = 3;
    private static final int DEFAULT_INTERVAL_SEC = 30;

    private SharedPreferences mPrefs;
    private TextView mDriftValue;
    private TextView mIntervalValue;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPrefs = getSharedPreferences(PREFS_NAME, MODE_WORLD_READABLE);

        ScrollView sv = new ScrollView(this);
        LinearLayout ll = new LinearLayout(this);
        ll.setOrientation(LinearLayout.VERTICAL);
        int pad = dp(24);
        ll.setPadding(pad, pad, pad, pad);

        // 标题
        TextView title = new TextView(this);
        title.setText("状态栏漂流瓶");
        title.setTextSize(24);
        title.setTextColor(Color.WHITE);
        title.setGravity(Gravity.CENTER_HORIZONTAL);
        ll.addView(title);

        TextView sub = new TextView(this);
        sub.setText("\n周期性水平偏移状态栏，防止 OLED 烧屏\n");
        sub.setTextSize(13);
        sub.setTextColor(0xFFB0B0B0);
        sub.setGravity(Gravity.CENTER_HORIZONTAL);
        ll.addView(sub);

        // 偏移幅度
        TextView driftLabel = new TextView(this);
        driftLabel.setText("偏移幅度");
        driftLabel.setTextSize(15);
        driftLabel.setTextColor(Color.WHITE);
        driftLabel.setPadding(0, dp(16), 0, dp(4));
        ll.addView(driftLabel);

        mDriftValue = new TextView(this);
        mDriftValue.setText(getCurrentDrift() + " px");
        mDriftValue.setTextSize(14);
        mDriftValue.setTextColor(0xFF4DD0E1);
        ll.addView(mDriftValue);

        SeekBar driftBar = new SeekBar(this);
        driftBar.setMax(10);
        driftBar.setProgress(getCurrentDrift() - 1);
        driftBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                int px = progress + 1;
                mDriftValue.setText(px + " px");
                if (fromUser) {
                    mPrefs.edit().putInt(KEY_DRIFT_PX, px).apply();
                }
            }
            @Override public void onStartTrackingTouch(SeekBar s) {}
            @Override public void onStopTrackingTouch(SeekBar s) {}
        });
        ll.addView(driftBar);

        // 切换周期
        TextView intervalLabel = new TextView(this);
        intervalLabel.setText("切换周期");
        intervalLabel.setTextSize(15);
        intervalLabel.setTextColor(Color.WHITE);
        intervalLabel.setPadding(0, dp(16), 0, dp(4));
        ll.addView(intervalLabel);

        mIntervalValue = new TextView(this);
        mIntervalValue.setText(getCurrentInterval() + " 秒");
        mIntervalValue.setTextSize(14);
        mIntervalValue.setTextColor(0xFF4DD0E1);
        ll.addView(mIntervalValue);

        SeekBar intervalBar = new SeekBar(this);
        intervalBar.setMax(115); // 5-120秒
        intervalBar.setProgress(getCurrentInterval() - 5);
        intervalBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                int sec = progress + 5;
                mIntervalValue.setText(sec + " 秒");
                if (fromUser) {
                    mPrefs.edit().putInt(KEY_INTERVAL_SEC, sec).apply();
                }
            }
            @Override public void onStartTrackingTouch(SeekBar s) {}
            @Override public void onStopTrackingTouch(SeekBar s) {}
        });
        ll.addView(intervalBar);

        // 重启 SystemUI 按钮
        Button restartBtn = new Button(this);
        restartBtn.setText("重启 SystemUI 立即生效");
        restartBtn.setTextColor(Color.WHITE);
        restartBtn.setBackgroundColor(0xFF00897B);
        restartBtn.setOnClickListener(v -> {
            try {
                Runtime.getRuntime().exec(new String[]{"su", "-c", "killall com.android.systemui"});
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
        LinearLayout.LayoutParams btnLp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, dp(48));
        btnLp.setMargins(0, dp(24), 0, 0);
        ll.addView(restartBtn, btnLp);

        // 说明
        TextView info = new TextView(this);
        info.setText(
            "\n\n【说明】\n" +
            "• 偏移幅度：状态栏左右漂移的最大距离\n" +
            "  1px = 几乎不可见，10px = 比较明显\n" +
            "  推荐 2-4px，防烧屏够用且不显眼\n\n" +
            "• 切换周期：每隔多久变一次位置\n" +
            "  周期越短防烧屏效果越好，但更耗电\n" +
            "  推荐 20-60 秒\n\n" +
            "修改后点「重启 SystemUI」立即生效\n\n" +
            "版本：1.2.0");
        info.setTextSize(12);
        info.setTextColor(0xFF909090);
        info.setLineSpacing(0, 1.3f);
        info.setPadding(0, dp(16), 0, 0);
        ll.addView(info);

        sv.addView(ll);
        sv.setBackgroundColor(0xFF121212);
        setContentView(sv);
    }

    private int getCurrentDrift() {
        return mPrefs.getInt(KEY_DRIFT_PX, DEFAULT_DRIFT_PX);
    }

    private int getCurrentInterval() {
        return mPrefs.getInt(KEY_INTERVAL_SEC, DEFAULT_INTERVAL_SEC);
    }

    private int dp(int px) {
        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, px,
                getResources().getDisplayMetrics());
    }
}

File: app/src/main/java/com/gtlx/statusbardrift/ui/MainActivity.kt
package com.gtlx.statusbardrift.ui

import android.app.Activity
import android.graphics.Color
import android.os.Bundle
import android.util.TypedValue
import android.view.Gravity
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.SeekBar
import android.util.Log
import android.widget.TextView
import com.gtlx.statusbardrift.config.DriftConfig

/**
 * 设置界面 —— 桌面入口
 *
 * 可调参数：
 * - 偏移幅度（1-10px）
 * - 切换周期（5-120 秒）
 * - 一键重启 SystemUI
 */
class MainActivity : Activity() {

    private var driftPx = 3
    private var intervalSec = 30

    private lateinit var driftValue: TextView
    private lateinit var intervalValue: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        loadConfig()
        // 启动时确保配置文件也写到 /data/local/tmp/，SystemUI 能读到
        syncConfigToTmp()


        val sv = ScrollView(this)
        val ll = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(24), dp(24), dp(24), dp(24))
        }

        // 标题
        ll.addView(TextView(this).apply {
            text = "状态栏漂流瓶"
            textSize = 24f
            setTextColor(Color.WHITE)
            gravity = Gravity.CENTER_HORIZONTAL
        })
        ll.addView(TextView(this).apply {
            text = "\n周期性水平偏移状态栏，防止 OLED 烧屏\n"
            textSize = 13f
            setTextColor(0xFFB0B0B0.toInt())
            gravity = Gravity.CENTER_HORIZONTAL
        })

        // 偏移幅度
        ll.addView(TextView(this).apply {
            text = "偏移幅度"
            textSize = 15f
            setTextColor(Color.WHITE)
            setPadding(0, dp(16), 0, dp(4))
        })
        driftValue = TextView(this).apply {
            text = "$driftPx px"
            textSize = 14f
            setTextColor(0xFF4DD0E1.toInt())
        }
        ll.addView(driftValue)
        ll.addView(SeekBar(this).apply {
            max = 10
            progress = driftPx - 1
            setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
                override fun onProgressChanged(sb: SeekBar, p: Int, fromUser: Boolean) {
                    driftPx = p + 1
                    driftValue.text = "$driftPx px"
                    if (fromUser) saveConfig()
                }
                override fun onStartTrackingTouch(sb: SeekBar) {}
                override fun onStopTrackingTouch(sb: SeekBar) {}
            })
        })

        // 切换周期
        ll.addView(TextView(this).apply {
            text = "切换周期"
            textSize = 15f
            setTextColor(Color.WHITE)
            setPadding(0, dp(16), 0, dp(4))
        })
        intervalValue = TextView(this).apply {
            text = "$intervalSec 秒"
            textSize = 14f
            setTextColor(0xFF4DD0E1.toInt())
        }
        ll.addView(intervalValue)
        ll.addView(SeekBar(this).apply {
            max = 115 // 5-120秒
            progress = intervalSec - 5
            setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
                override fun onProgressChanged(sb: SeekBar, p: Int, fromUser: Boolean) {
                    intervalSec = p + 5
                    intervalValue.text = "$intervalSec 秒"
                    if (fromUser) saveConfig()
                }
                override fun onStartTrackingTouch(sb: SeekBar) {}
                override fun onStopTrackingTouch(sb: SeekBar) {}
            })
        })

        // 重启 SystemUI 按钮
        val restartBtn = Button(this).apply {
            text = "重启 SystemUI 立即生效"
            setTextColor(Color.WHITE)
            setBackgroundColor(0xFF00897B.toInt())
            setOnClickListener {
                try {
                    Runtime.getRuntime().exec(arrayOf("su", "-c", "killall com.android.systemui"))
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
        ll.addView(restartBtn, LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT, dp(48)
        ).apply { setMargins(0, dp(24), 0, 0) })

        // 说明
        ll.addView(TextView(this).apply {
            text = """
                
                【说明】
                • 偏移幅度：状态栏左右漂移的最大距离
                  1px = 几乎不可见，10px = 比较明显
                  推荐 2-4px，防烧屏够用且不显眼

                • 切换周期：每隔多久变一次位置
                  周期越短防烧屏效果越好，但更耗电
                  推荐 20-60 秒

                修改后配置自动热更新，大多数情况无需重启
                （幅度变化立即生效，周期变化下次切换时生效）

                版本：1.0.0
            """.trimIndent()
            textSize = 12f
            setTextColor(0xFF909090.toInt())
            setLineSpacing(0f, 1.3f)
            setPadding(0, dp(16), 0, 0)
        })

        sv.addView(ll)
        sv.setBackgroundColor(0xFF121212.toInt())
        setContentView(sv)
    }

    private fun loadConfig() {
        // 从文件读（跟 SystemUI 读同一份）
        DriftConfig.load(this)
        driftPx = DriftConfig.driftPx
        intervalSec = (DriftConfig.intervalMs / 1000).toInt()
    }

    private fun saveConfig() {
        DriftConfig.saveFromUi(this, driftPx, intervalSec)
    }

    private fun syncConfigToTmp() {
        // 用 root 权限把配置文件复制到 /data/local/tmp/，SystemUI 能读到
        Thread {
            try {
                val dir = getExternalFilesDir(null) ?: filesDir
                val src = java.io.File(dir, "status_bar_drift.conf")
                if (!src.exists()) saveConfig()
                if (src.exists()) {
                    val proc = Runtime.getRuntime().exec(arrayOf(
                        "su", "-c",
                        "cp ${src.absolutePath} /data/local/tmp/status_bar_drift.conf && chmod 644 /data/local/tmp/status_bar_drift.conf"
                    ))
                    val exitCode = proc.waitFor()
                    Log.d("StatusBarDrift", "syncConfigToTmp exitCode=$exitCode")
                }
            } catch (t: Throwable) {
                Log.e("StatusBarDrift", "syncConfigToTmp failed", t)
            }
        }.start()
    }

    private fun dp(px: Int): Int = TypedValue.applyDimension(
        TypedValue.COMPLEX_UNIT_DIP, px.toFloat(),
        resources.displayMetrics
    ).toInt()
}
